### ISSUE TITLE
Trying to create a Windows 98 instance with latest code fails [macOS]

### ISSUE THREAD

**Issue by brunocastello** [joncampbell123/dosbox-x#1873]
**Describe the bug**
An error (attached below) is given while installing Windows 98 SE from CD-ROM method. DOSBox-X crashes on 1st reboot during install and shows the attached screen.

**To Reproduce**
Follow the steps for [Installing windows 98 in DOSBox-X](https://github.com/Wengier/dosbox-x-wiki/wiki/Guide%3AInstalling-Windows-98)

**Expected behavior**
A complete working Windows 98 install, as seen on previous build (0.83.5)

**Screenshots**
![Screen Shot 2020-09-21 at 6 52 09 PM](https://user-images.githubusercontent.com/23453202/93825829-691d8100-fc3c-11ea-88b9-a15ddf803ae0.png)

**Environment (please complete the following information):**
 - Operating system: macOS Catalina 10.15.6
 - DOSBox-X release version or commit SHA1 this issue occurred on: latest code as of today (Sep 21 2020).
 - Used configuration: same as seen in guide above

**Additional context**
None

**Comment by brunocastello** [comment]
I can continue the install when I use the previous build but then Windows 98 throws a lot of errors before completing the install. One on first reboot (WPF) another while detecting devices (garbled BSOD with no messages) and another when a screen to select date/time zone, closer to finishing the install (BSOD, fatal exception 06 has occurred at 001D:0000035B). 

I remember clearly that I had managed a clean install ONCE, without errors. I think the errors are thrown (one of them was a Windows Protection Error) because DOSBox-X is too fast even though I am not using turbo mode. I tried lowering the cycles to half (60000 to 30000) but no dice. I might try again with lower cycles.

**Comment by brunocastello** [comment]
I discovered the reason why it was crashing - nothing to do with the above.

When I switched output=opengl to output=surface, it booted. Why, what has changed in between builds with regards to OpenGL support?

**Comment by Wengier** [comment]
@brunocastello Support for OpenGL (GLSL) shaders was just added yesterday. So you are saying GLSL shade can cause crash? I will disable the built-in shader by default if this is the case, at least for now.

**Comment by brunocastello** [comment]
> @brunocastello Support for OpenGL (GLSL) shaders was just added yesterday. So you are saying OpenGL shaders can cause crash? I will disable the built-in shader by default if this is the case, at least for now.

@Wengier I cannot confirm this yet, because I suspect that the default value for output is 'default' and apparently DOSBox-X should guess which one based on host OS. I think it may be guessing wrong, and I was lucky it booted with output=surface, so I decided to experiment with output=openglnb to confirm this before answering your question. Right now I am going through the install process again. I'll let you know the results in around 25 minutes or less. 

If it fails, then it will be definitely the case, if not, then it may be the process how DOSBOX-X is guessing the default output? Does it make any sense?

**Comment by brunocastello** [comment]
@Wengier I can confirm that this is the issue. both output=opengl and output=openglnb crashes the Windows 98 SE install when it does its first reboot of the install process process. When I change to output=surface it boots.

**Comment by Wengier** [comment]
@brunocastello Can you check the latest code with the opengl output?

**Comment by brunocastello** [comment]
I’ll try that right now. It may take a while. Because trying turbo mode or high cycles apparently is causing protection errors on Windows 98 SE install process.

**Comment by brunocastello** [comment]
No luck, same errors. Must be something else. I'll try to investigate further in the next days.

**Comment by rderooy** [comment]
Also on Linux with either the SDL1 or SDL2 build, if you try to run Win98SE (did not try others), with output=opengl it will crash almost immediately after the boot command. All I get in the log is a Segmentation fault.

There is a slight difference between SDL1 and SDL2.
- SDL1 crashes instantly after the boot command, the last message on the screen is "Booting from drive C...."
- SDL2 crashes slightly later, in that it just manages to show the Win98SE boot splash screen

**Comment by Wengier** [comment]
@rderooy Can you post the config file and the full log, and if possible a test case for the issue to occur? I know it is very strange, but it indeed did not happen when I try to boot from a disk image (Windows 98 or PC DOS 2000). If it could be triggered here so easily I should have found and fixed it much earlier.

![image](https://user-images.githubusercontent.com/8216923/93927174-827b0780-fce6-11ea-8d54-0f6066a7e15c.png)

![image](https://user-images.githubusercontent.com/8216923/93928139-dafed480-fce7-11ea-9175-97436ee62d21.png)

The config file I used for the tests was based on the one provided by @brunocastello.

```
[sdl]
output=opengl
autolock=true

[dosbox]
title=Windows 98
memsize=128

[video]
vmemsize=4
vesa modelist width limit=0
vesa modelist height limit=0

[dos]
ver=7.1

[cpu]
cputype=pentium
core=normal
cycles=60000

[sblaster]
sbtype=sb16vibra
irq=5

[ne2000]
# If you want networking in Windows, set ne2000=true.
# This also requires that you set realnic= to a suitable value for your PC
ne2000  = true
nicbase = 300
nicirq  = 3
macaddr = AC:DE:48:88:99:DD
realnic = en4

[ide, primary]
enable=true
pnp=false
int13fakeio=true
int13fakev86io=true

[fdc, primary]
enable=true
pnp=false
int13fakev86io=true

[ide, secondary]
enable=true
pnp=true
int13fakeio=true
int13fakev86io=true
cd-rom insertion delay=4000

[render]
scaler=none

[autoexec]
imgmount d "d:/iso/Win95_OSR25.iso" -ide 2m
imgmount c "d:/imgs/win98.img" -reservecyl 1
boot c:
```

**Comment by rderooy** [comment]
Ok, so the PC DOS 2000 seems like it may be slightly different. It seems to be related somehow to the cpu core= type.

With the following config, booting PC-DOS 2000 works:
```
[autoexec]
mount c hdd.img
boot c:
```
This will use output=surface by default. But If I change it to output=opengl, it crashes with this as the last message.
```
LOG: Limit check 110000+4-1 = 110003 > ffff
LOG: Segment limit violation
LOG: DOSBox-X has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox-X's options.
LOG: Setting execute permission on the code cache has failed
Segmentation fault
```
I know this issue, it is https://github.com/dosbox-staging/dosbox-staging/issues/253 and it can be worked around by changing the core type. And indeed setting core=normal allows it to run.

But why is core=normal suddenly needed if output=opengl?

And also, my win98 config obviously has core=normal set as otherwise win9x has lots of problems.

**Comment by rderooy** [comment]
I tried stripping my win98.conf to the absolute bare minimum, and it still crashes with output=opengl but I can boot just fine from the Win98SE CD-ROM.

I will try to do some more testing tomorrow afternoon or friday, and get a more decent, hopefully reproducible test case.

**Comment by Wengier** [comment]
@rderooy I was not able to reproduce the crash with PC DOS 2000 with either cpu core= type. However, after a few tries I did find one thing that can trigger a crash with "output=opengl", that is, when I run and exit from Windows 3.1 on another computer of mine (which did not happen in my main computer), so I have already fixed this crash in the latest code. Please check if this also fixes the crash you encountered. Thanks.

**Comment by brunocastello** [comment]
Just to let you know that I still cannot test. A blackout of 25 hours is still going on so no internet and no power. I just borrowed a battery to charge my phone and come online for a few minutes to check out the things. I still cannot test the files from september 9, 2020 nor the latest changes.

If this helps, When I did the install process of Windows 98SE two days ago with the config file I provided, does not matter if I keep or remove the lines imgmount a -bootcd d and boot -l a on first OS boot during install, it still crashed with the popup message.

I basically started DOSBOX-X with the config file, went into Install, scandisk checks the disk, proceed to the start of the install, choose type of install (custom) and insert all the data it asks and product key. When the install finishes copying the files, it asks for a reboot, so I allow it to reboot and dang, crash. I can only continue if I choose output=surface, but the install (even if I start with output=surface) is damaged anyway because on second reboot after selecting date/time, it hangs with a Windows Protection Error. 10/10 times trying to install I get some BSOD or WPE. Actually, I do get a BSOD when I try to select my date/timezone and hit “apply”, and also when I get a window to select the correct IRQ for NE2000 (from IRQ3 to IRQ10).

On 0.83.5, I managed to do a clean install without errors only once. No errors during install and no problems in Device Manager in only one attempt. 9/10 times I would get errors during install. All that with the CD-ROM install method. If I do the second method, I get problems in Device Manager. I have not tried the second method with later versions.

A quick research suggests me that the errors during the install will happen if the emulation speed is too fast for it to handle. I could not test it properly because even going from 60000 cycles to 30000 cycles the errors would not disappear. I know it’s not my CD image the problem because I can install just fine with the same Windows 98SE CD image in QEMU, VMware and PCem.

I did attempt to install with core=auto and core=dynamic one or two times, changing it back to normal when I was asked to reboot into OS first time. Had mixed luck with it. I also tried cpu pentium and pentium_mmx.

**Comment by rderooy** [comment]
@Wengier I just compiled the latest code and gave it a quick try.

It did not change the PC DOS 2000 issue, which is selinux related with the dynamic core. But I was now able to boot Win98SE with output=opengl

**Comment by Wengier** [comment]
@rderooy Since booting PC DOS 2000 is never an issue on my computers, can you check if it works on your end before a specific commit?

**Comment by rderooy** [comment]
@Wengier the PC DOS 2000 issue is selinux stepping in and killing the process. Here is the selinux log about it:
```
SELinux is preventing dosbox-x-sdl2 from using the execheap access on a process.

*****  Plugin allow_execheap (53.1 confidence) suggests   ********************

If you do not think dosbox-x-sdl2 should need to map heap memory that is both writable and executable.
Then you need to report a bug. This is a potentially dangerous access.
Do
contact your security administrator and report this issue.

*****  Plugin catchall_boolean (42.6 confidence) suggests   ******************

If you want to allow selinuxuser to execheap
Then you must tell SELinux about this by enabling the 'selinuxuser_execheap' boolean.

Do
setsebool -P selinuxuser_execheap 1

*****  Plugin catchall (5.76 confidence) suggests   **************************

If you believe that dosbox-x-sdl2 should be allowed execheap access on processes labeled unconfined_t by default.
Then you should report this as a bug.
You can generate a local policy module to allow this access.
Do
allow this access for now by executing:
# ausearch -c 'dosbox-x-sdl2' --raw | audit2allow -M my-dosboxxsdl2
# semodule -X 300 -i my-dosboxxsdl2.pp

Additional Information:
Source Context                unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1
                              023
Target Context                unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1
                              023
Target Objects                Unknown [ process ]
Source                        dosbox-x-sdl2
```

In any case, I tried a few other DOS versions, booting them with output=opengl but *not* setting core=normal (leaving it to default) and after a bit of troubleshooting I was able to tune in to the cause.

The moment EMM386.EXE get's loaded in CONFIG.SYS the crash happens. I tried it both without parameters and with the NOEMS parameter and the effect was the same.

**Comment by Wengier** [comment]
@rderooy Does this happen all the time or only after a specific commit (e.g. after GLSL shader support was added)? I don't usually load EMM386 in the boot image, but I just tried to add EMM386 in both MS-DOS 7.1 and PC DOS 2000 and they still booted to the DOS prompt without issues. Tested with the settings "output=opengl" and "core=dynamic".

**Comment by brunocastello** [comment]
I'm back, after 41 hours of blackout and a lawsuit from my condo against the power company, I have power and internet again. What do I have to test? September 9, 2020 commit code to build and test, right?

**Comment by rderooy** [comment]
Before and after https://github.com/joncampbell123/dosbox-x/commit/1954d6bfba6dfbdfbd323c2eb9b2277548487d46 would be best.

**Comment by brunocastello** [comment]
I'm on my way, give me a few hours

**Comment by rderooy** [comment]
@Wengier I just quickly tried it on another machine running F32, where I still had the rather old v0.82.22 installed. And I did not run into the issue. Will try to do another test with a more recent version.

Also, what Linux distro are you using in your VM? Does it have selinux? Here is how you can check if selinux is installed, if it is active:
```
$ getenforce 
Enforcing
```

**Comment by rderooy** [comment]
With the latest DOSBox-X...
- With MS-DOS 6.22 it crashes like I mentioned on EMM386, but it also hangs when loading COUNTRY.SYS giving a never ending stream of
```
LOG:  206410634 ERROR CPU:Illegal Unhandled Interrupt Called 6
```
The COUNTRY.SYS problem is **not** solved by switching to core=normal

- With PC DOS 2000 it crashes on EMM386, but it also crashes on DISPLAY.SYS with this line:
```
DEVICE=C:\DOS\DISPLAY.SYS CON=(,,1)
```
COUNTRY.SYS however seems to load fine with PC DOS 2000.

A few other things I noticed. While my default dosbox-x config has core=auto, and typing ``config -get core`` returns ``auto``, the drop down CPU menu has no option for auto and instead claims I'm running the ``normal`` core.
So I tried forcing dynamic by setting ``core=dynamic`` in my config file, and I cannot even start dosbox-x any longer. I get an immediate segmentation fault with output=opengl combined with core=dynamic (no selinux error)
```
[sdl]
output=opengl
[cpu]
core=dynamic
```

I also build 0.83.5 and it runs the above tests just fine. So I will try to hone in on the offending commit.

**Comment by rderooy** [comment]
I did a git bisect on it, and as suspected, this is the offending commit:
```
$ git bisect bad
a01ac1c1680ea93368e24cd63fba38ba8c0ee747 is the first bad commit
commit a01ac1c1680ea93368e24cd63fba38ba8c0ee747
Author: Wengier <*******@yahoo.com>
Date:   Sun Sep 20 17:48:33 2020 -0400

    support for GLSL shader on both SDL1 and SDL2 builds

 CHANGELOG                             |   7 +
 contrib/linux/dosbox-x.appdata.xml.in |   2 +-
 dosbox-x.reference.conf               |   5 +
 include/build_timestamp.h             |   4 +-
 include/render.h                      |   6 +-
 include/video.h                       |   2 +-
 src/dosbox.cpp                        |   8 +
 src/gui/render.cpp                    |  92 +++++-
 src/gui/render_glsl.h                 | 576 ++++++++++++++++++++++++++++++++++
 src/gui/sdlmain.cpp                   | 142 ++++++++-
 src/output/output_opengl.cpp          | 286 +++++++++++++++--
 src/output/output_opengl.h            |  54 ++++
 12 files changed, 1146 insertions(+), 38 deletions(-)
 create mode 100644 src/gui/render_glsl.h
```

**Comment by Wengier** [comment]
@rderooy I am running Fedora 30, and SELinux seems to be active here as well. Here is the screenshot:

![image](https://user-images.githubusercontent.com/8216923/94148598-d13eb380-fe44-11ea-84f9-6b6842e26a39.png)

**Comment by Wengier** [comment]
@rderooy Since no crash with MS-DOS or PC DOS 2000 with the opengl setting here at all, can you please try the following sdlmain.cpp file to replace the current one? This will disable the GLSL shader initialization process:

[sdlmain.zip](https://github.com/joncampbell123/dosbox-x/files/5276311/sdlmain.zip)

Let me know if it works on your end.

**Comment by brunocastello** [comment]
Windows 98 is a hit and miss to install on DOSBox, more than before now. I've been trying to complete the install (CD-ROM method) but I'm unable to reach the end with so many errors, none of them related to opengl. Last night I've tried like ten times to install with the code from before the 1954d6b 

I have a few ideas regarding the conf file, but I'll have to try later because now I have to work. All I can say is that there a few settings from the configuration file that need to be disabled or enabled for Win 9x, all them are described in the conf file itself.

But I'm afraid that I will have to resort to the 2nd method (local install) and try to fix the other issues post-install later.

**Comment by Wengier** [comment]
@brunocastello I think the code from before the commit 1954d6b has nothing to do with the Windows 98 installation. I believe @rderooy asked you try the older commit because he wants to check if the following warning you mentioned in the other thread still occurs:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/objdump: error: 'dosbox-x.app/Contents/MacOS/*.dylib': No such file or directory
```

Try the latest code for the Windows 98 installation.

**Comment by brunocastello** [comment]
> @brunocastello I think the code from before the commit [1954d6b](https://github.com/joncampbell123/dosbox-x/commit/1954d6bfba6dfbdfbd323c2eb9b2277548487d46) has nothing to do with the Windows 98 installation. I believe @rderooy asked you try the older commit because he wants to check if the following warning you mentioned in the other thread (#1871) still occurs:
> 
> ```
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/objdump:
> error: 'dosbox-x.app/Contents/MacOS/*.dylib': No such file or directory
> ```
> 
> Please check if the above warning is still there with the older commit. On the other hand, for the Windows 98 installation please try the latest code.

Yes, the warning still occurs with the code from before the commit [1954d6b](https://github.com/joncampbell123/dosbox-x/commit/1954d6bfba6dfbdfbd323c2eb9b2277548487d46). Just tried it now.

As for Windows 98. I'll try later the latest code.

**Comment by Wengier** [comment]
@brunocastello Just to make sure, can you also try the commit https://github.com/joncampbell123/dosbox-x/commit/39ef11f7de3a6eb42294ab121ba3f849fa52325a to see if the warning still occurs? It is before the file movements started.

**Comment by rderooy** [comment]
@Wengier I can confirm that replacing sdlmain.cpp and rebuilding prevents the crash.

**Comment by Wengier** [comment]
@rderooy So the code that causes crash on your system is between line 3909 and 3962 of sdlmain.cpp. I can comment out these lines as in this test version, but this will completely disable GLSL shader support, which is not desired. So I want to know exactly which line(s) may cause the problem on your system. For this purpose I have uploaded 3 *slightly* different versions of sdlmain.cpp (named sdlmain1.cpp, sdlmain2.cpp and sdlmain3.cpp) which comment out different parts of that code. Please check which one(s) may work on your system. Thanks!

[sdlmain.zip](https://github.com/joncampbell123/dosbox-x/files/5277573/sdlmain.zip)

**Comment by rderooy** [comment]
@Wengier Just did 3 builds, and all three crashed.

**Comment by Wengier** [comment]
@rderooy Are you using SDL1 or SDL2 build?

**Comment by rderooy** [comment]
@wengier in this case for all 4 builds, I did SDL1. Let me know if you want me to do SDL2 also.

**Comment by Wengier** [comment]
@rderooy Can you try again with the following, which again includes 3 slightly different versions of sdlmain.cpp (named sdlmain1.cpp, sdlmain2.cpp and sdlmain3.cpp):

[sdlmain.zip](https://github.com/joncampbell123/dosbox-x/files/5277836/sdlmain.zip)

Let me know which one(s) may work on your system (SDL1 build). Thanks!

**Comment by rderooy** [comment]
- sdlmain1.cpp works
- sdlmain2.cpp crashes
- sdlmain3.cpp works

**Comment by brunocastello** [comment]
> @brunocastello Just to make sure, can you also try the commit [39ef11f](https://github.com/joncampbell123/dosbox-x/commit/39ef11f7de3a6eb42294ab121ba3f849fa52325a) to see if the warning still occurs? It is before the file movements started.

Yes, it still occurs. I compile "./build-macosx" (always for SDL1, because SDL2 has issues with yellow tinted screen on Windows 3.11 if I use higher color resolutions, but SDL1 does not - original DOSBox has the same issue though) and then I issue "make dosbox-x.app", it's when the aforementioned error/warning regarding the dylibs happens.

I've been trying to install Windows 98 SE on latest code and so far no luck, the install crashes all the time, I can't blame opengl because I don't even get to a point where I can boot to the desktop and finish the install - which is when the opengl problems started days ago. It's hit and miss. crashes when the install gets to detect and set up the pnp devices, and crashes after I select and apply my region date/timezone - which is the last thing to do before rebooting into desktop. And eventually on 1st reboot during install I get a Windows Protection Error. I'm sure I've seen that error before in my life - I've done countless of Windows 98 reinstalls when something broke decades ago when I was young and had an Acer Extensa 710T. So my memories tells me that this happens because DOSBox is going too fast for the OS. I tried to make it slower, chose pentium over pentium_mmx, reduced cycles to 25000 or 30000, but it's not helping too.

I am doing one last attempt with the latest code build right now and if it fails, and I'm going to revert to 0.83.5 and try one last time before giving up and doing the 2nd method of install - local install from C:\WIN98 after copying the files. I just hope I can get the Device Manager issues fixed later. I did have a working Windows 98 install at least three or four times between these compiled builds weeks ago.

**Comment by Wengier** [comment]
@rderooy Thanks! Can you try one last time that if the following two slightly different versions of sdlmain.cpp (named sdlmain1.cpp, sdlmain2.cpp) work on your system?

[sdlmain.zip](https://github.com/joncampbell123/dosbox-x/files/5278219/sdlmain.zip)

**Comment by Wengier** [comment]
@brunocastello If the build you tried 3 days ago when you made this thread can boot to the desktop, then the latest build should certainly be able to do it as well, because in the default setting (i.e. unless you set mididevice=mt32) there should be virtually no difference between these two builds. Please check if anything has changed in your environment.

**Comment by Wengier** [comment]
@brunocastello Also, since the warning occurred even in commit 39ef11f as you said, it means the warning is not relevant to recent file movements at all. It is likely due to some change in your environment as well, or that it simply occurs at the time.

**Comment by Wengier** [comment]
@brunocastello Moreover, you mentioned that the code before the commit 1954d6b gave you errors as well just like today's tests, even though with the intermediate commit (3 days ago) you can boot to the desktop. This must mean there are some change(s) in your environment which gave you such mixed results.

**Comment by brunocastello** [comment]
I'm sorry, but there are no environment changes on my side. I've also tested on another (rarely used) iMac, same results.

These dylibs were always missing from all the builds I've done so far, and are also missing from the latest builds available for download on releases page.

The 0.83.5 build is the only one I have that can boot (to complete the install) with output=opengl. But installing Windows 98 by CD-ROM method without errors is a hit and miss game on DOSBox-X, and I've only managed that twice. I can continue the install with the errors but then APM and also CD-ROM will not work, so I cannot finish the install on last step or shutdown/reboot properly.

Later builds only boot to desktop (to complete the install) with output=surface, but they also suffer from the same Windows 98 install errors as above.

I use the same install media on other emulators (PCem and QEMU) as well as VMware and they never crashed.

**Comment by rderooy** [comment]
@Wengier 

- sdlmain1.cpp works
- sdlmain2.cpp works

**Comment by Wengier** [comment]
@brunocastello So you are saying that installing Windows 98 by CD-ROM method is always a hit and miss, which is the case for 0.85.3 build also. That is why the Wiki recommends users to copy the files to a disk image first, which is not really difficult and can save you from a lot troubles. As first the OpenGL issue, I think it is already solved in the latest code. Why not check it out and install Windows 98 by copying the files from the CD-ROM to the disk image first?

**Comment by Wengier** [comment]
@rderooy Great! I have already updated the file. I will consider it solved if it works fine on your system.

**Comment by brunocastello** [comment]
> @brunocastello So you are saying that installing Windows 98 by CD-ROM method is always a hit and miss, which is the case for 0.85.3 build also. That is why the Wiki recommends users to copy the files to a disk image first, which is not really difficult and can save you from a lot troubles. As for the OpenGL issue, I think it is already solved in the latest code. Why not check it out and install Windows 98 by copying the files from the CD-ROM to the disk image first (not the CD-ROM method)?

The problem with the second method is that after the install, the CD-ROM drive do not work (so I cannot install software or games), and APM never works (so I cannot shutdown or reboot Windows 98, I'm forced to do hard reset or close DOSBox-X).

the first method, when it works and does a clean install, do not show these problems at all. But only when it works.

I can and will try again later the second method, but I do not expect a better result. I'll be back later. I almost managed another CD-ROM clean install on 0.83.5 without blue screens or protection errors right now, I just failed near the end on desktop, when Windows 98 asked for the CD-ROM and couldn't find it.

**Comment by rderooy** [comment]
@brunocastello the APM problems where due to setting different APM settings than the reference config. You had PnP enabled, which is even specifically mentioned in the reference config not to do as it causes problems, and you also changed the default APM version, which I suspect may be another cause of problems.

The issues you had with IDE I don't understand. But note that if Win9x complains that the controller is in 16bit mode, is not a real problem. There is in my measurements no real performance benefit when run in DOSBox-X if the IDE controller is in 16 or 32bit mode as both are emulated paths.

**Comment by rderooy** [comment]
@Wengier I can confirm that the SDL1 code now works, but the SDL2 code path still crashes.

**Comment by brunocastello** [comment]
> @brunocastello the APM problems where due to setting different APM settings than the reference config. You had PnP enabled, which is even specifically mentioned in the reference config not to do as it causes problems, and you also changed the default APM version, which I suspect may be another cause of problems.
> 

Even when I tried to install without changes to the APM settings the problem still arises. So my changes haven't caused the problem, they were my reaction to try and fix that problem. Actually, when I managed the two successful installs, I had this:

`apmbios = true`
`apmbios pnp = false`
`apmbios version = 1.2`
`apmbios allow realmode = true`
`apmbios allow 16-bit protected mode = false`
`apmbios allow 32-bit protected mode = true`

I had Windows 98 with APM working and was able to reboot and shutdown without problems. But only if I manage to install without errors. Changing these with the second method install did not help though. Also, the two install methods had different estimated sizes: 308MB vs 306MB, which leads me to suggest that something is missing.

> The issues you had with IDE I don't understand. But note that if Win9x complains that the controller is in 16bit mode, is not a real problem. There is in my measurements no real performance benefit when run in DOSBox-X if the IDE controller is in 16 or 32bit mode as both are emulated paths.

I believe that the el torito method is the culprit for the CD-ROM problems I'm having here. If I can change the region date/timezone while installing and without a BSOD when I confirm the change, I know the install will for sure proceed without errors, so I can just quit DOSBox before the next reboot, remove the method on config file and change it to boot from the hard disk, then CD-ROM will be detected with all my absolute sure. I've seen that happen in these two successful installs so I am gonna try it now.

**Comment by Wengier** [comment]
@rderooy SDL1 and SDL2 use somewhat different code in a few places. Can you check if the following two slightly different versions of sdlmain.cpp (sdlmain1.cpp and sdlmain2.cpp) work on your system for SDL2? Thanks.

[sdlmain.zip](https://github.com/joncampbell123/dosbox-x/files/5279041/sdlmain.zip)

**Comment by rderooy** [comment]
@Wengier 

- sdlmain1.cpp crashes
- sdlmain2.cpp crashes

**Comment by Wengier** [comment]
@rderooy I think you only tried my previously-uploaded sdlmain.cpp files with SDL1. Can you try them with SDL2 this time and report the results? Thanks.

**Comment by Wengier** [comment]
@rderooy You can first test the one I posted first (i.e. [sdlmain.zip](https://github.com/joncampbell123/dosbox-x/files/5276311/sdlmain.zip)), which disables the GL shader initialization process completely. If it does not work either, then there is no need to check for the others.

**Comment by rderooy** [comment]
@Wengier 

This gets a little confusing since the files are all called the same. I assume you mean the first one that contained 3 variants in the zip file?

- sdlmain1.cpp crashes
- sdlmain2.cpp crashes
- sdlmain3.cpp crashes

**update**
I just noticed your later comment. With that sdlmain.cpp it also crashes.

**Comment by Wengier** [comment]
@rderooy Disabling the GL shader initialization process completely still crashes for SDL2 on your system? Are you sure the SDL2 crash only occurs since GL shader support was added? Also, have you checked the log file and see anything abnormal in it?

**Comment by brunocastello** [comment]
I finally was able to complete the installation on 0.83.5 with the CD-ROM method, resorting to something I did not want to do and was suggested by @rderooy on another issue.

Basically, I configured the cd-rom on autoexec.bat and config.sys (using oakcdrom.sys) before the last reboot into desktop and then booted into. Once I was on desktop, however, Windows was not able to detect the installation media, it asked for the CD-ROM media and I had to point to D:\WIN98 when asked. I completed the install and went to Device Manager expecting to see more trouble. I had only two IDE controllers, one with an exclamation mark (which I removed before shutting down). 

I expect to be able to remove the workaround after I edit a few lines from my config file, because I never had to do a such thing before and because after that the CD-ROM was still readable. Let's see. It worked for me before. If all goes well again, then I will compile another DOSBox-X build with the latest code available and verify if the existing install still works on the new build.

I'll confirm that tomorrow because now it's midnight and I need to sleep before another day at work.

**Comment by brunocastello** [comment]
I actually sucessfully compiled a 0.83.6 build and booted into the newly existing Windows 98 SE image. All working now - APM, reboot, shutdown, CD-ROM, Floppy and IDE controllers. I'll just copy the WIN98 folder from CD media to C:\ and edit the registry to point the the installation media to that folder whenever it asks for that. Damn it, almost one week struggling.

@Wengier I did not have any issue to boot into the system with output=opengl this time around. But I installed the system with the 0.83.5, confirmed it working, then I shutdown, compiled 0.83.6 and booted into desktop minutes later with 0.83.6 and output=opengl. I am not sure if the issue will remain if someone decides to install Windows 98, but with the existing install, it works.

Next days I have to update the stuff and find out if I can run FIFA 98 or 99 on it, but performance wise needs a lot of work from previous tests - Cycles, CPU, Core, needs tweaks. Last time I tested FIFA 99 weeks ago, it had artifacts and crashed.
